### DISCUSSION TITLE
wix extensions not resolved correctly in msbuild

### DISCUSSION THREAD

**Discussion by gkellerer** [wixtoolset/issues#7989] [Questions] [answered]
Hello,

i am faced with a problem a can't get a solution for some time and hope that somebody can help me. I tried to build a wix bundle in visual studio with an msbuild .wixproj project. I added the following references to my project file:

```xml
  <ItemGroup>
    <PackageReference Include="WixToolset.Bal.wixext" Version="4.0.4" />
    <PackageReference Include="WixToolset.BalUtil" Version="4.0.4" />
    <PackageReference Include="WixToolset.Heat" Version="4.0.4" />
    <PackageReference Include="WixToolset.Util.wixext" Version="4.0.4" />
  </ItemGroup>
```

my bundle.wxs file contains the following lines:

```xml
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs"
     xmlns:bal="http://wixtoolset.org/schemas/v4/wxs/bal"
     xmlns:util="http://wixtoolset.org/schemas/v4/wxs/util">
  <Bundle DisableModify="yes" Name="Infowas Service V$(var.ProductVersion)" Version="$(var.ProductVersion)" Manufacturer="LfU" UpgradeCode="61444485-14b3-4765-a103-f4c703f67eba">
    <BootstrapperApplication>
      <bal:WixStandardBootstrapperApplication LicenseUrl="" Theme="hyperlinkLicense" />
    </BootstrapperApplication>
...
```

when i compile it i get an wix0200 error indicating that elements from teh bal namespace are not present. the references are shown correctly in solution explorer under visual studio. but when i check the "installed" tab on the nuget packages it shows empty even i installed the extensions before. during build a saw that the wixextensions are not resolved correctly. Can anyone tell me what i could try else? I am using Visual Studio 2022 pro (17.8.6) under windows 11 with wix 4.0.4.

thank you very much,
Georg
6

**Comment by robmen** [comment]
I would build from the command-line using `msbuild -Restore -bl` to download the package referenced and create a log file. Then I'd look in the log for the command-line for `wix.exe` to see if the extension really is being pass along. The log file will explain a lot.

**Comment by gkellerer** [comment]
Hello,

when building from the command line the extensions are found and used with the -ext parameter. When I start the build process from within visual studio the -ext parameter is missing completely in the wix command line.

Von: Rob Mensching ***@***.***>
Gesendet: Freitag, 9. Februar 2024 15:38
An: wixtoolset/issues ***@***.***>
Cc: Kellerer, Georg (LfU) ***@***.***>; Author ***@***.***>
Betreff: Re: [wixtoolset/issues] wix extensions not resolved correctly in msbuild (Discussion #7989)


I would build from the command-line using msbuild -Restore -bl to download the package referenced and create a log file. Then I'd look in the log for the command-line for wix.exe to see if the extension really is being pass along. The log file will explain a lot.

**Comment by gkellerer** [comment]
in the output window of visual studio i saw
[...]
ResolveWixExtensionReferences;
[...]
the target "ResolveWixExtensionReferences" has been skipped because the condition was "false". ( '@(WixExtension)' != '') has been evaluated to ( '' != '').
[...]
(translated from a german vs version). but i still couldn't find out where something may be missing

**Reply by robmen** [reply]
That suggests your projects are not being restored. `PackageReference`s are downloaded from NuGet so maybe you've done something on your system to prevent that from working.

**Reply by joshprittse** [reply]
I was getting this exact error with our old .net framework builds on jenkins and Wix 6.0.  It was giving the error: 
error WIX0094: The identifier 'Dialog:ErrorDlg' could not be found. Ensure you have typed the reference correctly and that all the necessary inputs are provided to the linker.
This went away and built just fine after we added the -Restore switch to the call to msbuild.exe in our Jenkinsfile.  Seems like nuget.exe restore does not restore these wix packages, only msbuild -Restore will do it if you're building old .net framework code.  I bet dotnet build would work just fine if we were building new .net core apps...

**Comment by gkellerer** [comment, marked as answer]
the following snippet in my .wixproj file solved my problem. Maybe there's a better solution especially for the HintPath but it works:
  <ItemGroup>
    <WixExtension Include="WixToolset.Bal.wixext" Version="4.0.4">
      <Name>WixToolset.Bal.wixext</Name>
      <HintPath>$(NuGetPackageRoot)\wixtoolset.bal.wixext\4.0.4\wixext4\WixToolset.Bal.wixext.dll</HintPath>
    </WixExtension>
    <WixExtension Include="wixtoolset.util.wixext" Version="4.0.4">
      <Name>WixToolset.Util.wixext</Name>
      <HintPath>$(NuGetPackageRoot)\wixtoolset.util.wixext\4.0.4\wixext4\WixToolset.util.wixext.dll</HintPath>
    </WixExtension>
  </ItemGroup>

**Reply by robmen** [reply]
That shouldn't be necessary, but if you're satisfied, go wild with it.
